### PR TITLE
Communities: add requests menu tab

### PR DESCRIPTION
**Pull request by Pineirin**
Fix internationalization bugs in the view.
closes inveniosoftware/invenio-communities#454
Screenshot from 2022-03-16 16-34-49

### CODE REVIEW

File: invenio_communities/assets/semantic-ui/js/invenio_communities/requests/index.js [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+      "receiver",
+      "community:" + community.uuid,
+    ]);
+    for (const [componentId, component] of Object.entries(
+      RequestsDefaultComponents
+    )) {
+      overrideStore.add(componentId, component);
+    }
+    return (
+      <>
```
Comment (kpsherva) [inline invenio_communities/assets/semantic-ui/js/invenio_communities/requests/index.js:41]: redundant React Fragment

File: invenio_communities/searchapp.py [hunk 2]
```diff
@@ -0,0 +1,29 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright (C) 2018-2021 CERN.
```
Comment (kpsherva) [inline invenio_communities/searchapp.py:3]:
```suggestion
# Copyright (C) 2018-2022 CERN.
```

File: invenio_communities/searchapp.py [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+
+def search_app_context():
+    """Search app context processor."""
+    return {
+        'search_community_requests_config': partial(
+            search_app_config,
+            'COMMUNITIES_SEARCH_REQUESTS',
+            current_app.config['REQUESTS_FACETS'],
+            current_app.config['COMMUNITIES_SORT_OPTIONS'],
+            '/api/requests',
```
Comment (kpsherva) [inline invenio_communities/searchapp.py:25]: hardcoded URL, please check with Javier if it can be avoided (he has the same config)

File: invenio_communities/templates/semantic-ui/invenio_communities/details/header.html [hunk 4]
```diff
@@ -73,7 +73,8 @@ <h2 class="ui header">{{ community.metadata.title }}</h2>
     </div>
     {% set header_menu_items = {
       'search': (_('Search'), url_for('invenio_communities.communities_detail', pid_value=community.id), 'search icon', True),
-      'members': (_('Members'), url_for('invenio_communities.members', pid_value=community.id), 'users icon', true),
+      'members': (_('Members'), url_for('invenio_communities.members', pid_value=community.id), 'users icon', True),
+      'requests': (_('Requests'), url_for('invenio_communities.communities_requests', pid_value=community.id), 'file outline icon', permissions.can_create),
```
Comment (kpsherva) [inline invenio_communities/templates/semantic-ui/invenio_communities/details/header.html:77]:
```suggestion
      'requests': (_('Requests'), url_for('invenio_communities.communities_requests', pid_value=community.id), 'comments outline icon', permissions.can_create),
```
what about comments icon instead to mark requests ?

File: invenio_communities/templates/semantic-ui/invenio_communities/details/requests/index.html [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+{%- set title = community.metadata.title + _(" requests") -%}
+
+{%- block javascript %}
+{{ super() }}
+{{ webpack['invenio-communities-requests.js'] }}
+{%- endblock javascript %}
+
+{%- block page_body %}
+{{ super() }}
+<div class="ui container grid">
```
Comment (kpsherva) [inline invenio_communities/templates/semantic-ui/invenio_communities/details/requests/index.html:21]: I am not sure why grid is needed here ?

File: invenio_communities/views/communities.py [hunk 6]
```diff
@@ -98,10 +98,32 @@ def communities_settings(community=None, logo=None, pid_value=None):
...
     )
 
+@require_create_permission
+@pass_community
+@pass_community_logo
+def communities_requests(community=None, logo=None, pid_value=None):
+    """Community requests page."""
+    return render_template(
+        "invenio_communities/details/requests/index.html",
+        community=community.to_dict(), # TODO: use serializer,
```
Comment (kpsherva) [inline invenio_communities/views/communities.py:112]: if community UI serializer is missing, could you create an issue for this?

File: invenio_communities/views/communities.py [hunk 7]
```diff
@@ -172,6 +194,14 @@ def invitations(community=None, pid_value=None):
...
+        permissions=permissions(community),
         active_menu_tab="members",
     )
+
+
+def permissions(community):
+    """Calculates permissions for the given community"""
+    permissions = {}
+    permissions.update(community.has_permissions_to(['update']))
+    permissions.update(community.has_permissions_to(['create']))
```
Comment (kpsherva) [inline invenio_communities/views/communities.py:206]: why is create permission needed?

File: invenio_communities/templates/semantic-ui/invenio_communities/details/requests/index.html [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+  {%- block requests_body %}
+  <div
+    id="app"
+    data-community='{{ community | tojson }}'
+    data-invenio-search-community-requests-config='{{ search_community_requests_config(app_id="user-requests-search") | tojson }}'>
+  </div>
+  <div class="ui hidden divider"></div>
```
Comment (kpsherva) [inline invenio_communities/templates/semantic-ui/invenio_communities/details/requests/index.html:28]: why do we need the divider?

File: invenio_communities/assets/semantic-ui/js/invenio_communities/requests/index.js [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+      _camelCase("invenio-search-community-requests-config")
+    );
+    config["initialQueryState"]["filters"].push([
+      "receiver",
+      "community:" + community.uuid,
+    ]);
+    for (const [componentId, component] of Object.entries(
+      RequestsDefaultComponents
+    )) {
+      overrideStore.add(componentId, component);
```
Comment (kpsherva) [inline invenio_communities/assets/semantic-ui/js/invenio_communities/requests/index.js:38]: why is this needed?

Comment (Pineirin) [reply invenio_communities/assets/semantic-ui/js/invenio_communities/requests/index.js:38]: As it was done for the dashboard, this piece of code overrides the content that will be displayed by the SearchApp component. In this case, the overrides of the components are defined here: https://github.com/inveniosoftware/invenio-communities/pull/455/files#diff-6662c287617406713a1a6b49085ecdf7fccf1e16f2b54c2635b574ff3be8684fR527-R540.

File: invenio_communities/assets/semantic-ui/js/invenio_communities/requests/index.js [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+
+const getConfigFromDataAttribute = (element, attr) => {
+  const dataValue = domContainer.dataset[attr];
+
+  return JSON.parse(dataValue);
+};
+
+class CommunityRequests extends Component {
+  render() {
+    const { appId, ...config } = getConfigFromDataAttribute(
```
Comment (ntarocco) [inline invenio_communities/assets/semantic-ui/js/invenio_communities/requests/index.js:27]: This should be moved in the constructor, and not re-evaluated on each `render`

File: invenio_communities/views/communities.py (hunk 7 above)
Comment (Pineirin) [reply invenio_communities/views/communities.py:206]: As far as I understand, the users that can see and access the requests tab are the users that have the can_create permissions for the community:
can_create = [AuthenticatedUser(), SystemProcess()]

In order to hide the tab from user that don't have the can_create permission, as it was done for the can_update permission with the settings, this information must be passed to the jinja template.

File: invenio_communities/config.py [hunk 11]
```diff
@@ -65,3 +66,37 @@
     ),
 }
 """Community roles."""
+
+COMMUNITIES_SEARCH_REQUESTS = {
```
Comment (ntarocco) [inline invenio_communities/config.py:70]: We have to come up with good name spacing here because there will be 4/5 search pages.
Can you please prepare a list of all the search pages and propose namespacing? Then, we can discuss it together.

File: invenio_communities/searchapp.py (hunk 3 above)
Comment (ntarocco) [reply invenio_communities/searchapp.py:25]: It can be injected from the `routes` config, as you have added it above.

File: invenio_communities/searchapp.py [hunk 12]
```diff
@@ -0,0 +1,29 @@
...
+            headers={"Accept": "application/json"},
+            initial_filters=[["is_open", "true"]]
```
Comment (ntarocco) [inline invenio_communities/searchapp.py:27]: It might be the place where you can inject here the community UUID

File: invenio_communities/assets/semantic-ui/js/invenio_communities/requests/index.js [hunk 13]
```diff
@@ -0,0 +1,49 @@
...
+      domContainer,
+      _camelCase("invenio-search-community-requests-config")
+    );
+    config["initialQueryState"]["filters"].push([
+      "receiver",
+      "community:" + community.uuid,
```
Comment (ntarocco) [inline invenio_communities/assets/semantic-ui/js/invenio_communities/requests/index.js:33]: This should be moved to backend? See comment below.

File: invenio_communities/templates/semantic-ui/invenio_communities/details/requests/index.html [hunk 14]
```diff
@@ -0,0 +1,33 @@
...
+{%- endblock javascript %}
+
+{%- block page_body %}
+{{ super() }}
+<div class="ui container grid">
+  {%- block requests_body %}
+  <div
+    id="app"
+    data-community='{{ community | tojson }}'
+    data-invenio-search-community-requests-config='{{ search_community_requests_config(app_id="user-requests-search") | tojson }}'>
```
Comment (ntarocco) [inline invenio_communities/templates/semantic-ui/invenio_communities/details/requests/index.html:26]: what about simply:
```suggestion
    data-config='{{ search_community_requests_config(app_id="user-requests-search") | tojson }}'>
```

File: invenio_communities/searchapp.py [hunk 15]
```diff
@@ -0,0 +1,29 @@
...
+
+from flask import current_app
+
+from invenio_search_ui.searchconfig import search_app_config
+
+
+def search_app_context():
+    """Search app context processor."""
+    return {
+        'search_community_requests_config': partial(
```
Comment (ntarocco) [inline invenio_communities/searchapp.py:20]: rework naming namespacing here

File: invenio_communities/views/decorators.py [hunk 16]
```diff
@@ -65,4 +65,20 @@ def view(**kwargs):
...
 
-    return view
\ No newline at end of file
+    return view
+
+def require_create_permission(f):
+    """Decorate a view to require community owner for accessing the view."""
+    @wraps(f)
+    def view(**kwargs):
+        """."""
```
Comment (ntarocco) [inline invenio_communities/views/decorators.py:74]: Missing docstring

File: invenio_communities/views/communities.py (hunk 7 above)
Comment (ntarocco) [reply invenio_communities/views/communities.py:206]: Actually, why are we touching permissions?

File: invenio_communities/views/communities.py [hunk 17]
```diff
@@ -98,10 +98,32 @@ def communities_settings(community=None, logo=None, pid_value=None):
         },
         # Pass permissions so we can disable partially UI components
         # e.g Settings tab
-        permissions=community.has_permissions_to(['update']),
+        permissions=permissions(community),
         active_menu_tab="settings"
     )
 
+@require_create_permission
```
Comment (ntarocco) [inline invenio_communities/views/communities.py:105]: I don't think this is correct: the tab should probably displayed when you are authenticated

File: invenio_communities/templates/semantic-ui/invenio_communities/details/header.html (hunk 4 above)
Comment (ntarocco) [inline invenio_communities/templates/semantic-ui/invenio_communities/details/header.html:77]: `permissions` to be reviewed and probably changed

File: invenio_communities/config.py (hunk 11 above)
Comment (Pineirin) [reply invenio_communities/config.py:70]: Communities records = COMMUNITIES_SEARCH_RECORDS
All communities = COMMUNITIES_SEARCH
Community requests = COMMUNITIES_SEARCH_REQUESTS
Community members = COMMUNITIES_SEARCH_MEMBERS
Community invitations = COMMUNITIES_SEARCH_INVITATIONS

File: invenio_communities/assets/semantic-ui/js/invenio_communities/requests/requests.js [hunk 18]
```diff
@@ -0,0 +1,536 @@
...
+      </Segment>
+    </>
+  );
+};
+
+export const CommunityRequestsEmptyResultsWithState = withState(
+  CommunityRequestsEmptyResults
+);
+
+export const defaultComponents = {
```
Comment (ntarocco) [inline invenio_communities/assets/semantic-ui/js/invenio_communities/requests/requests.js:523]: Please remove all the `export` above and you can also avoid to namespace them with `Community...`

```suggestion
export const defaultComponents = (appId) => {
  return {
      buildUID('SearchApp.facets', '', appId): RequestsFacets,
      ...
  }
```

File: invenio_communities/assets/semantic-ui/js/invenio_communities/requests/requests.js [hunk 19]
```diff
@@ -0,0 +1,536 @@
...
+                Biodiversity Literature Repository
+              </span>
+            </>
+          )}
+        </Item.Meta>
+      </Item.Content>
+    </Item>
+  );
+}
+// FIXME: Keeping ResultsGrid.item and SearchBar.element because otherwise
```
Comment (ntarocco) [inline invenio_communities/assets/semantic-ui/js/invenio_communities/requests/requests.js:322]: Can you have a look to this and remove unused components?

**Review by ntarocco — APPROVED** [review]

**Review by zzacharo — APPROVED** [review]
